src/libaktualizr/uptane/managedsecondary.cc

```diff
@@ -122,17 +136,20 @@ bool ManagedSecondary::sendFirmware(const std::string &data) {
       if (boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha256digest(data))) !=
           boost::algorithm::to_lower_copy(it->HashString())) {
         detected_attack = "wrong_hash";
+        *events_channel << std::make_shared<event::InstallComplete>(getSerial());
         return true;
       }
     } else if (it->TypeString() == "sha512") {
       if (boost::algorithm::to_lower_copy(boost::algorithm::hex(Crypto::sha512digest(data))) !=
           boost::algorithm::to_lower_copy(it->HashString())) {
         detected_attack = "wrong_hash";
+        *events_channel << std::make_shared<event::InstallComplete>(getSerial());
         return true;
       }
     }
   }
   detected_attack = "";
+  *events_channel << std::make_shared<event::InstallComplete>(getSerial());
```

Comment:
Sorry, I think I'm missing something here. Why are we sending `InstallComplete` before `storeFirmware` is called? And why are we sending `InstalllComplete` at all in any of the error cases?

Reply:
About 'InstallComplete before storeFirmware' yes you are right. 
About 'InstalllComplete' at error. It is because at error we cannot continue with install, so error means that installation is really finished, but with error. And when all ecus finished, error will be included in the ecus manifest.